msprime/trees.py

```diff
@@ -2235,14 +2434,17 @@ def branch_stats_vector(self, leaf_sets, weight_fun, windows=None):
         and averages this across the tree sequence, weighted by genomic length.
 
         It does this separately for each window [windows[i], windows[i+1])
-        and returns the values in a vector.
+        and returns the values in a list.
         '''
         if windows is None:
             windows = (0, self.sequence_length)
         for U in leaf_sets:
             if len(U) != len(set(U)):
                 raise ValueError(
                     "elements of leaf_sets cannot contain repeated elements.")
+            for u in U:
+                assert(type(u) is int)
```

Comment:
This looks more like an input check than a sanity check to me... The first check is actively bad I think, as it violates Python rules about allowing duck-typing. Is there any particular reason that it must be precisely an int?

The second check should be changed to a ValueError I think. 

My rule of thumb with asserts vs exceptions is, asserts should only be raised when a mistake happens _within_ msprime, not when the _user_ makes a mistake. I.e., asserts are just for sanity checking, not input validation. Bad inputs supplied by the user should always raise an exception.